### PR TITLE
Fixes for handling _DEFAULT tenant in NotConfigured mode

### LINKED ISSUE #1123: [Multi-ESX] admin cli vm-group access ls command fails with default configuration
- No configuration is set on the esx:

```
[root@sc-rdops-vm04-dhcp-106-152:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group create --name=tenant1
Internal Error(Error: Please init configuration in vmdkops_admin before trying to change it)

```
- Only have _DEFAULT tenant on the ESX
```
[root@sc-rdops-vm04-dhcp-106-152:~] usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group ls
Uuid                                  Name      Description                 Default_datastore  VM_list
------------------------------------  --------  --------------------------  -----------------  -------
11111111-1111-1111-1111-111111111111  _DEFAULT  This is a default vm-group

```
- Try running ```admin cli vm-group access ls``` command but see error stacktrace.

```
[root@sc-rdops-vm04-dhcp-106-152:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group access ls --name=_DEFAULT
Traceback (most recent call last):
  File "/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py", line 1356, in <module>
    main()
  File "/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py", line 58, in main
    args.func(args)
  File "/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py", line 1121, in tenant_access_ls
    rows = generate_tenant_access_ls_rows(privileges)
  File "/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py", line 1100, in generate_tenant_access_ls_rows
    if not p.datastore_url or p.datastore_url == auth_data_const.DEFAULT_DS_URL:
AttributeError: 'dict' object has no attribute 'datastore_url'
[root@sc-rdops-vm04-dhcp-106-152:~]

```

### PR DESCRIPTION
**Pull request by msterin**
Fixes #1123 

Fixes for handling _DEFAULT tenant in NotConfigured mode

The majority of functions in auth_api.py are decorated with @onlywhenconfigured.
Some of them are not :
(1) when internal check for vmgroup name allows _DEFAULT
(2) when these are internal function and supposed to be called only via API

Also use proper error handling for multi-node related code in auth_api.c
with errinfo/codes and the likes

Test: manual:
```
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup update --name=_DEFAULT --description="New description of default tenant"
Please init configuration in vmdkops_admin before trying to change it
[root@localhost:~]  /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup create --name=tenant1
Please init configuration in vmdkops_admin before trying to change it
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup update --name=_DEFAULT --new-name=new-vm-group1
Please init configuration in vmdkops_admin before trying to change it
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup update --name=_DEFAULT --default-datastore=sharedVmfs-0
Please init configuration in vmdkops_admin before trying to change it
vmdkops_admin.py vmgroup access add: error: the following arguments are required: --datastore
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup access add --name=_DEFAULT --datastore=Datastore1
Please init configuration in vmdkops_admin before trying to change it
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group access add --name=_DEFAULT
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup access rm --name=_DEFAULT --datastore=Datastore1
Please init configuration in vmdkops_admin before trying to change it
[root@localhost:~] usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup access set --name=_DEFAULT --datastore=Datastore1 --allow-create=False
Please init configuration in vmdkops_admin before trying to change it
[root@localhost:~]  /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup vm add --name=tenant-t1 --vm-list=photon-VM0.2
Please init configuration in vmdkops_admin before trying to change it
```

### CODE REVIEW

File: esx_service/utils/auth_api.py [hunk 1]
```diff
@@ -43,6 +45,53 @@ def get_auth_mgr_object():
         return error_info, None
     return None, auth_mgr
 
+
+def only_when_configured(ret_obj=False):
+    """
+    Decorator to check if the DB was already inited.
+    Serves functions which return ErrInfo (when ret_obj=False) ,and the ones
```
Comment (shaominchen) [inline esx_service/utils/auth_api.py]: Nit: move the space before ","

File: esx_service/utils/auth_api.py [hunk 2]
```diff
@@ -43,6 +45,53 @@ def get_auth_mgr_object():
...
+    returning (ErrInfo, None) when ret_obj=True.
+    Makes sure the decorated function is called only when DB is connected,
+    otherwise a proper ErrInfo is returned.
+    """
+
+    def real_decorator(func):
+        'The actual logic for decorator.'
+
+        def not_inited(*_, **__):
+            'Returns err code for not initited'
```
Comment (shaominchen) [inline esx_service/utils/auth_api.py]: Nit: initited => inited or initialized

File: esx_service/utils/auth_api.py [hunk 3]
```diff
@@ -43,6 +45,53 @@ def get_auth_mgr_object():
...
+            return error_code.generate_error_info(ErrorCode.INIT_NEEDED)
+
+        def internal_error(*_, **__):
+            'Rerturns error code for internal errors'
+            return error_code.generate_error_info(ErrorCode.INTERNAL_ERROR,
+                                                  "@only_when_configured: %s" % func.__name__)
+
+        def check_config(*args, **kwargs):
+            'call func() if DB is configured and issue an error it not.'
```
Comment (shaominchen) [inline esx_service/utils/auth_api.py]: Nit: it => if

File: esx_service/utils/error_code.py [hunk 4]
```diff
@@ -46,11 +46,18 @@ class ErrorCode:
     # DATASTORE related error code start
     DEFAULT_DS_NOT_SET = 301
     DS_NOT_EXIST = 302
+
+    # indicates that config init needed for a requested operation
```
Comment (shaominchen) [inline esx_service/utils/error_code.py]: Nit: it seems some of our comments start with capital letter, some don't. Let's keep them consistent.

Comment (msterin) [inline esx_service/utils/error_code.py]: @shaominchen - thanks for the review ! All fixed, will update the PR shortly

File: esx_service/utils/auth_api.py [hunk 5]
```diff
@@ -43,6 +45,53 @@ def get_auth_mgr_object():
...
+
+    def real_decorator(func):
+        'The actual logic for decorator.'
+
+        def not_inited():
+            'Returns err code for not initialized'
+            return error_code.generate_error_info(ErrorCode.INIT_NEEDED)
+
+        def internal_error():
+            'Rerturns error code for internal errors'
```
Comment (pdhamdhere) [inline esx_service/utils/auth_api.py]: Typo: Rerturns

File: esx_service/utils/error_code.py [hunk 6]
```diff
@@ -86,6 +93,8 @@ class ErrorCode:
 
     ErrorCode.DEFAULT_DS_NOT_SET : "Default datastore is not set",
     ErrorCode.DS_NOT_EXIST : "Datastore {0} does not exist",
+    ErrorCode.INIT_NEEDED: "Please init configuration in vmdkops_admin before trying to change it",
```
Comment (pdhamdhere) [inline esx_service/utils/error_code.py]: Even better: "Please init configuration using 'vmdkops_admin config init' before trying to change it"

Comment (msterin) [inline esx_service/utils/error_code.py]: thanks. fixed and merged

**Pull request merged**